beaker-ts/src/notebook.ts

```diff
@@ -558,43 +586,118 @@ export class BeakerQueryCell extends BeakerBaseCell implements IQueryCell {
+
+        // only tag first cell with all metadata to reconstruct, and let the rest get dropped on fromIPynb
+        let parentCellPushed = false;
+        const parentMetadata = {
+            ...this.metadata,
+            beaker_cell_type: "query",
+            prompt: this.source,
+            events: this.events,
+        };
+        const openIndentSection = `
+<div style="
+display: flex;
+width: 100%;
+">
+<div style="
+border: 1px solid var(--jp-cell-editor-border-color) !important;
+border-radius: var(--jp-border-radius);
+padding: 0.2rem;
+margin: 0.2rem;
+margin-left: 4rem;
+">
+`
+        const closeIndentSection = `</div></div>`
+
+        // add first User: tagged section identical to how the agent ones are.
+        let markdownLinesBuffer = [
+            `### User:\n`,
+            openIndentSection,
+            `${this.source}\n`
+        ]
+        let cells: nbformat.IBaseCell[] = [];
+
+        // make markdown cell from current buffer of lines from the agent output and flush it
+        // so that the current contents are all written before a code cell or other type is written
+        const pushNewAgentMarkdownCell = (additionalTags?: object) => {
+            markdownLinesBuffer.push(closeIndentSection);
+            const renderedMarkdown = markdownLinesBuffer.join("\n");
+            markdownLinesBuffer = [
+                "### Agent:",
+                openIndentSection
+            ];
+
+            if (parentCellPushed) {
+                // redundant information to parent, so ignore loading
+                cells.push(new BeakerMarkdownCell(
+                    {
+                        cell_type: "markdown",
+                        id: uuidv4(),
+                        source: renderedMarkdown,
+                        metadata: { skipWhenLoading: true, beakerQueryCellChild: true, ...(additionalTags ?? {}) },
+                    }
+                ));
+            }
+            else {
+                cells.push(new BeakerMarkdownCell(
+                    {
+                        cell_type: "markdown",
+                        id: this.id,
+                        source: renderedMarkdown,
+                        // contains all parent metadata
+                        metadata: { ...parentMetadata, parentQueryCell: true, beakerQueryCellChild: true, ...(additionalTags ?? {}) },
+                    }
+                ));
+                parentCellPushed = true;
+            };
+        }
+        // push user cell before the first agent thought
+        pushNewAgentMarkdownCell();
+
+        this.events.forEach(event => {
             if (event.type === "thought") {
-                renderedMarkdownLines.push(`> Thought: ${event.content.thought}\n> `);
+                markdownLinesBuffer.push(`${event.content.thought}  \n`);
+            }
+            else if (event.type === "user_question") {
+                markdownLinesBuffer.push(`**Clarification:**  \n> ${event.content}\n`);
```

Comment:
Not all questions are asking for clarification. It's probably fine, but perhaps would be better to just be "Agent question" or something more generic.